### PR TITLE
Matching-service communicates with collab-service when 2 users match

### PR DESCRIPTION
**Pull request by zacwong2151**
sorry for long commit :/

## TLDR: what happens when 2 users match

1. `matching-service`  sends info to  `collab-service` (through kafka)
2. `collab-service`  sends info to  `frontend` (thru API call)

## slightly more details

1. now when `matching-service` matches 2 users, it sends an event to `collab-service` thru kafka

2. `collab-service` listens to the event, then handles the logic of choosing a question from the database, given the topics and difficulties. Once done `collab-service` puts the information in a local data structure called `collabStore`. `collabStore` is a single source of truth that holds the information required for collaboration, for all users. At the same time, `frontend` navigates to `collab-page`, and requests the following information from `collab-service` through an API call:

    - the person that you matched with's user ID -> `matchedUserId`
    - `roomId`
    - the selected `questionId`

3. frontend uses `matchedUserId` and `questionId` to retrieve matched user's name and question details. Done

## TODO:
1. in `collab-service/collabController.ts`, there is a TODO There
2. ~~in `CollaborationPage`, there is a TODO there~~ **done**
3. ~~When users 'end session', remove them from userStore -> need to implement a 'end session' button~~ **done**
4. ~~When user attempts to enter the collaboration page, verify his ID against `collabStore`~~ **done**
5. When users 'log out', remove them from userStore

### CODE REVIEW

**Comment by brein62** [comment]
at least most of your long commit is in `package-lock.json` lmao

File: compose.yaml [hunk 1]
```diff
@@ -76,7 +76,8 @@ services:
       - ./collaboration-service:/usr/src/collaboration-service
       - /usr/src/collaboration-service/node_modules
     ports:
-      - "${COLLABORATIONSVC_PORT:-3001}:3001"
+      - "${COLLABORATIONSVC_PORT:-3001}:3001" # WebSocket server
+      - "3002:3002"  # REST API server
```
Comment (brein62) [inline compose.yaml:80]: change this line to maintain consistency with the other docker compose ports: 
`"${COLLABORATIONSVCRESTAPI_PORT:-3002}:3002" # REST API server`

File: frontend/src/api/collaboration-service/CollaborationService.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+const COLLABORATION_SERVICE_URL = "http://localhost:3002/"
+const COLLABORATION_BASE_URL = "/collaboration"
+const REMOVE_USER_URL = "/remove"
+
+const api = axios.create({
+    baseURL : COLLABORATION_SERVICE_URL,
+    withCredentials : true
+})
+
+async function getCollaborationInformation(userId : string) {
```
Comment (brein62) [inline frontend/src/api/collaboration-service/CollaborationService.ts:12]: `getCollabInfo` sounds more succinct but ok hehe

File: collaboration-service/src/utils/userStore.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { User } from "../models/user";
+
+// UserStore is a singleton class. Users will be added to this UserStore ONLY when they are collaborating
+// TODO: when users 'end session', remove them from this UserStore
```
Comment (brein62) [inline collaboration-service/src/utils/userStore.ts:4]: seems like you've already done this so you can remove this comment

File: matching-service/src/controllers/matchingController.ts [hunk 4]
```diff
@@ -117,10 +118,12 @@ export const startMatching = async () => {
                 matchedUser.matchedUser = newUser;
                 newUser.matchedUser = matchedUser;
 
-                // give both users a room Id to collaborate
-                const roomId = uuidv4()
-                matchedUser.roomId = roomId 
-                newUser.roomId = roomId 
+                const roomId = uuidv4() // give both users a room Id to collaborate 
+                const question_topics = findCommonTopics(newUser, matchedUser)
+                const question_difficulties = findCommonDifficulties(newUser, matchedUser)
```
Comment (brein62) [inline matching-service/src/controllers/matchingController.ts:123]: is this to be implemented by max?

File: matching-service/src/model/user.ts [hunk 5]
```diff
@@ -9,7 +9,19 @@ interface User {
...
     isPeerReady : boolean;
     matchedUser : User | null;
-    roomId: string | null; // uuid for collaboration-service
+}
+
+const findCommonTopics = (user1: User, user2: User) : string[] => {
+    return user1.topics.filter(topic => user2.topics.includes(topic))
+}
+
+// repeated code here, can refactor next time?
```
Comment (brein62) [inline matching-service/src/model/user.ts:18]: nah it's fine, the code is quite short

**Review by brein62 — COMMENTED** [review]
just some small feedback from a cursory look at the code

File: compose.yaml (hunk 1 above)
Comment (zacwong2151) [reply compose.yaml]: oki done

File: frontend/src/api/collaboration-service/CollaborationService.ts (hunk 2 above)
Comment (zacwong2151) [reply frontend/src/api/collaboration-service/CollaborationService.ts]: sounds good

File: collaboration-service/src/utils/collabStore.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    printContents() {
+        const contents = Array.from(this.collabStore.entries()).map(([key, value]) => ({
+            userId: key,
+            matchedUserId: value.matchedUserId,
+            roomId: value.roomId,
+            questionId: value.questionId,
+        }));
+        
+        console.table(contents);
+    }
```
Comment (wjayee) [inline collaboration-service/src/utils/collabStore.ts:37]: if ure using this to log who is currently in the collab store it may be bttr to only print either the user id so itll be less cluttered in the console but either way works

File: frontend/src/pages/matching-service/WaitForMatchingPage.tsx [hunk 7]
```diff
@@ -43,40 +44,49 @@ export default function WaitForMatchingPage() {
...
-          //Do nothing
-          console.log("matching...");
+
+    // Send request to check matching state
+    const response = await sendCheckMatchingStateRequest(auth.id);
+    
+    if (response.status === 200) {
+        console.log('2 users are matched')
+        // Check if the required information is in collaboration-service, before navigating to collab page
+        const response = await isUserInCollabStore(auth.id);
```
Comment (wjayee) [inline frontend/src/pages/matching-service/WaitForMatchingPage.tsx:63]: i think this logic should be in matching not collab so if a user is already collabing, the user should not be matched in the first place

Comment (wjayee) [reply frontend/src/pages/matching-service/WaitForMatchingPage.tsx:63]: thats why i mentioned collab service sending matching service the updated list of users collabing so that when first starting to match users, matching service will compare the user against that list to avoid this scenario

**Review by wjayee — CHANGES_REQUESTED** [review]
js a few comments

File: frontend/src/pages/collaboration-service/CollaborationPage.tsx [hunk 8]
```diff
@@ -1,22 +1,99 @@
...
+          // Means that user is not in user store, so he cannot access the collab-page
+          navigate("/matching/start")
+        }
+      } catch (error) {
+        console.error(error)
+      }
+    }
+
+    checkIfUserInStore()
+  }, [])
```
Comment (tiongMax) [inline frontend/src/pages/collaboration-service/CollaborationPage.tsx:53]: (Continue from 1) 2. I am thinking of storing the question information in the response of collab store then render in frontend, instead of sending another request from frontend to question service. What do u think? (Feel free to comment for the rest!)

**Review by tiongMax — CHANGES_REQUESTED** [review]
Just some thoughts.
